core/src/types/storage.rs

```diff
@@ -1151,6 +1154,147 @@ pub struct PrefixValue {
     pub value: Vec<u8>,
 }
 
+/// Container of all Ethereum event queues.
+#[derive(Default, Debug, BorshSerialize, BorshDeserialize)]
+pub struct EthEventsQueue {
+    /// Queue of transfer to Namada events.
+    pub transfers_to_namada: InnerEthEventsQueue<TransfersToNamada>,
+    // TODO: add queue of update whitelist events
+}
+
+/// A queue of confirmed Ethereum events to be processed in order.
+#[derive(Debug, BorshSerialize, BorshDeserialize)]
+pub struct InnerEthEventsQueue<E> {
+    next_nonce_to_process: Uint,
+    inner: VecDeque<E>,
+}
+
+impl<E: GetEventNonce> InnerEthEventsQueue<E> {
+    /// Return an Ethereum events queue starting at the specified nonce.
+    pub fn new_at(next_nonce_to_process: Uint) -> Self {
+        Self {
+            next_nonce_to_process,
+            ..Default::default()
+        }
+    }
+}
+
+impl<E: GetEventNonce> Default for InnerEthEventsQueue<E> {
+    fn default() -> Self {
+        Self {
+            next_nonce_to_process: 0u64.into(),
+            inner: Default::default(),
+        }
+    }
+}
+
+/// Draining iterator over a queue of Ethereum events.
+pub struct EthEventsQueueIter<'queue, E> {
+    current_nonce: Uint,
+    queue: &'queue mut InnerEthEventsQueue<E>,
+}
+
+impl<E> Drop for EthEventsQueueIter<'_, E> {
+    fn drop(&mut self) {
+        // on drop, we commit the nonce of the next event to process
+        if self.queue.next_nonce_to_process < self.current_nonce {
+            self.queue.next_nonce_to_process = self.current_nonce;
+        }
+    }
+}
+
+impl<E: GetEventNonce> Iterator for EthEventsQueueIter<'_, E> {
+    type Item = E;
+
+    fn next(&mut self) -> Option<E> {
+        let nonce_in_queue = self.queue.peek_event_nonce()?;
+        if nonce_in_queue == self.current_nonce {
+            self.current_nonce = self
+                .current_nonce
+                .checked_increment()
+                .expect("Nonce overflow");
+            self.queue.pop_event()
+        } else {
+            None
+        }
+    }
+}
+
+impl<E: GetEventNonce> InnerEthEventsQueue<E> {
+    /// Retrieve the next events to be processed, if any.
+    ///
+    /// This decision is based on the nonce of the next event to be
+    /// processed by the ledger, and the nonce of the event that was just
+    /// confirmed (i.e. achieved a quorum of votes behind it).
+    pub fn get_next_events(
```

Comment:
I find this method name confusing. Maybe something like `push_and_iter`?  Or `push_and_next`?